README.md
# Power consumption logger for Cmatic


Log data from cmatic measurement unit


# Get started

1. Checkout at github:
2. cd into the project directory
3. run the preinstall
```
# install prerequisites and the module itself
# make the command available
pip install -e .
```
4. Understand the logger by calling

```
pcl --help
```
5. Test the connection
```
pcl --host 'cmatic-abcdef'
```
where abcdef is replaced by your device


HISTORY.rst
=======
History
=======

0.1.0 (2020-08-14)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for power_consumption_logger."""


tests/test_power_consumption_logger.py
#!/usr/bin/env python

"""Tests for `power_consumption_logger` package."""

import pytest
import tempfile
import os
import responses
import requests


import power_consumption_logger.power_consumption_logger as pcl


@pytest.fixture(scope="module")
def temp_dir ():
    with tempfile.TemporaryDirectory() as tmpdirname:
        yield tmpdirname  # provide the fixture value


def helper_entry_available(filename, entry, line_no):
    with open(filename) as f:
        content = f.read().splitlines()
        print(content)
        return content[line_no] == pcl.data_to_str(entry)

def test_create_storage_dir__and_fail():
    """ we do not have create access rights on / root dir so it fails"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        pcl.Writer('/Z:/', 'simulate')
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 2

def test_create_storage_dir__ok(temp_dir):
    """ must be able to create something inside temp dir """
    new_dir = os.path.join(temp_dir, 'new')
    sut = pcl.Writer(new_dir, 'simulate')
    assert sut.dirname == new_dir
    assert os.path.isdir(new_dir)

def test_create_new_measurement_file__one_line(temp_dir):
    data = ['2020-08-19T22:55:00', 1, 1, 1, 1]
    sut = pcl.Writer(temp_dir, 'simulate')
    expected_file = os.path.join(temp_dir, 'cmatic_2020-08-19.csv')
    sut._write_data(data)
    assert os.path.isfile(expected_file)
    assert helper_entry_available(expected_file, data, 0)

def test_create_new_measurement_file__three_line(temp_dir):
    data1 = ['2020-08-20T08:55:00', 1, 1, 1, 1]
    data2 = ['2020-08-20T08:56:00', 1, 1, 1, 1]
    data3 = ['2020-08-20T08:57:00', 1, 1, 1, 1]
    sut = pcl.Writer(temp_dir, 'simulate')
    expected_file = os.path.join(temp_dir, 'cmatic_2020-08-20.csv')
    sut._write_data(data1)
    sut._write_data(data2)
    sut._write_data(data3)
    assert os.path.isfile(expected_file)
    assert helper_entry_available(expected_file, data1, 0)
    assert helper_entry_available(expected_file, data2, 1)
    assert helper_entry_available(expected_file, data3, 2)


@responses.activate
def test_web_response__ok():
    responses.add(responses.GET, 'http://cmatic-d9cd14/cm?cmnd=Status%2010',
        json={
            "StatusSNS": {
                "Time": "2018-11-11T13:46:56",
                "COUNTER": {
                    "C1": 109,
                    "C2": 8,
                    "C3": 6,
                "C4": 14
                }
            }
        }, status=200)

    expected_response = ['2018-11-11T13:46:56',  109, 8, 6, 14]
    assert pcl.read_from_web('D9CD14') == expected_response

@responses.activate
def test_web_response__connection_error():
    """ we do not have create access rights on / root dir so it fails"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        pcl.read_from_web('D9CD14')
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 1


power_consumption_logger/cli.py
"""Console script for power_consumption_logger."""
import argparse
import sys
import power_consumption_logger.power_consumption_logger as pcl


def main():
    """Console script for power_consumption_logger."""
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", default='cmatic-D9CD14',
        help="""The the hostname: if not set cmatic-D9CD14 is used
            if host is set to 'simulate' artificial values (0.1)
            are delivered without invoking the web service
            """)
    parser.add_argument("--storage", default='.',
        help="Set a storage directory: if not set current directory is used")
    parser.add_argument("-p", "--period", type=int,
        help="""Period where the readout happens (in seconds),
            if not set the program reads once and writes to console
            """)
    args = parser.parse_args()
    if args.period == None:
        # one time readout for testing
        print('# timestamp C1   C2   C3   C4')
        print(data_to_str(data))
    else:
        # read continuously and do not return
        recorder = pcl.Writer(args.storage, args.host)
        recorder.record(args.period)
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


power_consumption_logger/power_consumption_logger.py
"""Main module."""
import requests
import datetime
import time
import sys
from pathlib import Path
import os


def read_from_web_sim():
    timestamp = datetime.datetime.now().isoformat(timespec='seconds')
    return [ timestamp, 1, 2, 3, 4]


def read_from_web(host='cmatic-xxxxx'):
    """Read the data from the host interface

    Reads a JSON struct like {
        "StatusSNS": {
            "Time": "2018-11-11T13:46:56",
            "COUNTER": {
                "C1": 109,
                "C2": 8,
                "C3": 6,
                "C4": 14
                }
            }
        }
    from a web CDN

    Parameters
    ----------

    Returns
    -------

    array of [timestamp: string, c1: int c2: int, c3: int, c4: int]


    Exception:
    ----------
    in case of read error it exits the program with code 1

    """
    if host == 'simulate':
        # shortcut for testing
        return read_from_web_sim()
    try:
        url = 'http://' + host + '/cm'
        # param = {'cmnd': 'Status 10'.encode("utf-8")}
        param = '?cmnd=Status%2010'
        print ('{} param {}'.format(url, param))
        r = requests.get(url + param)
        data = r.json()
        print(data)
        return [
            data['StatusSNS']['Time'],
            data['StatusSNS']['COUNTER']['C1'],
            data['StatusSNS']['COUNTER']['C2'],
            data['StatusSNS']['COUNTER']['C3'],
            data['StatusSNS']['COUNTER']['C4']
        ]
    except requests.exceptions.ConnectionError as err:
        print('Connection error: {}'.format(err))
        sys.exit(1)


def data_to_str(data):
    return '{};{};{};{};{}'.format(data[0], data[1], data[2], data[3], data[4])

class Writer:

    def __init__(self, dirname, host):
        self.dirname = dirname
        self.host = host
        # make sure file exists and is writeable or throw an io exception
        try:
            path=Path(self.dirname)
            path.mkdir(parents=True, exist_ok=True)
        except EnvironmentError:
            # parent of IOError, OSError *and* WindowsError where available
            print('Cannot create storage location {}'.format(self.dirname))
            sys.exit(2)

    def _write_data(self, data):
        """Write data to file persistently.

        TODO add more

        Parameters:
        -----------

        data: structure like:
        "StatusSNS": {
            "Time": "2018-11-11T13:46:56",
            "COUNTER": {
                "C1": 109,
                "C2": 8,
                "C3": 6,
                "C4": 14
                }
            }
        }


        """
        iso_date = data[0][:10]  # 1st 10 letters form iso format are the date
        filename = os.path.join(self.dirname, 'cmatic_' + iso_date + '.csv')

        try:
            with open(filename, 'a') as f:
                f.write(data_to_str(data) + '\n')
        except EnvironmentError:
            # parent of IOError, OSError *and* WindowsError where available
            print('Cannot write data to file {}'.format(file_name))
            sys.exit(2)

    def record(self, period_in_sec: int):
        """Record a dataset continuously

        In a single thread, block while sleeping
        Never return
        """
        while True:
            self._write_data(read_from_web(self.host))
            time.sleep(period_in_sec)
